Enable Security's Cypress tests on all PRs

oatkiller opened the pull request:
## Summary

Security Solution writes e2e and other tests using Cypress. In the past, these tests, if they failed on a tracked branch, couldn't be easily skipped. They also weren't run in parallelized jobs. For primarily these reasons, they didn't run on most Kibana PRs. 

This PR moves these Cypress tests back to the main PR pipeline. Tests that fail on tracked branches create (or update) Github issues which can be used with the skip-test github workflow script to easily skip the failing tests. The pipeline steps are parallelized and run in under 40 minutes.

### Open Questions

- [x] Should this PR enable Serverless Security Defend Workflows Cypress Tests @patrykkopycinski 

### Some buildkite pipelines that used to run only on Security PRs now run on all PRs:

These steps run on all PRs with these changes

- Security Solution Cypress Tests (general tests that haven't been organized into an area team)
- Explore tests
- Investigations Tests
- Defend Workflows Tests
- Defend Workflows Serverless
- Threat Intel Tests
- OS Query Tests
- Security Solution Burning Changed Specs (these run only recently changed specs a few extra times)
- Security Solution OpenAPI codegen
- OSQuery burning
- OSQuery Serverless

<details>
  <summary><b>And these already run on all PRs</b></summary>

  - Serverless Security Cypress Tests
  - Serverless Explore tests
  - Serverless Investigations Tests
</details>

### Security Cypress tests run in the main `on merge` pipeline instead of the `on merge unsupported ftrs` pipeline:

These steps run in the `on merge` pipeline with these changes:

- Security Solution Cypress Tests
- Explore Cypress Tests
- Investigations Cypress Tests
- Defend Workflows Cypress Tests
- Defend Workflows Serverless Cypress Tests
- Threat Intelligence Cypress Tests
- Osquery Cypress Tests

<details>
  <summary><b>and these already run on the `on merge` pipeline</b></summary>

- Serverless Security Cypress Tests
- Serverless Explore - Security Solution Cypress Tests
- Serverless Investigations - Security Solution Cypress Tests

</details>

### Additional work to be done:

We need to consolidate build steps, enhance test skipping to support Cypress-grep flags, avoid out-of-memory errors in cypress, enhance parallelization, improve Cypress reporting, and probably other things. These are tracked separately. Reach out to me if you need details.

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

oatkiller commented on .buildkite/pipelines/pull_request/security_solution.yml:
```diff
@@ -1,40 +1,4 @@
 steps:
-  - command: .buildkite/scripts/steps/functional/security_solution.sh
```
These were moved to the pull_request base.yml so they will always run.

oatkiller commented on .buildkite/pipelines/pull_request/osquery_cypress.yml:
```diff
@@ -1,18 +1,4 @@
 steps:
-  - command: .buildkite/scripts/steps/functional/osquery_cypress.sh
```
This was moved to the base.yml so it will always run.

oatkiller commented on .buildkite/pipelines/pull_request/defend_workflows.yml:
```diff
@@ -1,26 +0,0 @@
-steps:
```
This was all moved to base.yml

oatkiller commented on .buildkite/pipelines/pull_request/base.yml:
```diff
@@ -80,6 +80,30 @@ steps:
   #       - exit_status: '*'
   #         limit: 1
 
+  - command: .buildkite/scripts/steps/functional/defend_workflows.sh
```
Defend Workflows Cypress tests, including serverless tests, will run on all PRs now.

jbudz commented:
:partying_face: !! How do we feel about including these in the on-merge pipeline too?  By expanding the scope to all pull requests, operations will typically be the contact for triage.  Our notifications come through via on-merge -  too much noise on pull requests.

MadameSheema commented:
> 🥳 !! How do we feel about including these in the on-merge pipeline too? By expanding the scope to all pull requests, operations will typically be the contact for triage. Our notifications come through via on-merge - too much noise on pull requests.

Completely agree with this!! @oatkiller when doing that change, please remember to remove our tests from `.buildkite/pipelines/on_merge_unsupported_ftrs.yml`

MadameSheema commented on .buildkite/scripts/pipelines/pull_request/pipeline.ts:
@oatkiller I believe we should do the same for the `threat intelligence` cypress tests. 

@PhilippeOberti may you please confirm if those tests are still owned by us? Thanks!

PhilippeOberti replied:
> @oatkiller I believe we should do the same for the `threat intelligence` cypress tests.
> 
> @PhilippeOberti may you please confirm if those tests are still owned by us? Thanks!

I'd say they are yes, as @lgestc and I are the ones who worked on them, and we're both in the Investigations' team

MadameSheema commented:
> These steps still only run on Security PRs
> 
> OS Query Cypress burns (these steps run new or changed tests multiple times before merge to ensure stability)
> General Security Solution Cypress burns (these steps run new or changed tests multiple times before merge to ensure stability)

@oatkiller what is the reason behind not moving these as well? I believe it would be good to have them as well, since if I'm not mistaken, are just triggered when there is a change in one of the cypress specs, that would help us a lot to control/surface flakiness and they are not blockers since are soft failures.

oatkiller commented:
> 🥳 !! How do we feel about including these in the on-merge pipeline too? By expanding the scope to all pull requests, operations will typically be the contact for triage. Our notifications come through via on-merge - too much noise on pull requests.

Sounds great, I will do that.

FYI: Currently these run in the 'on merge unsupported ftrs' pipeline. That pipeline now has the environment variable set to report failed test as github issues.

oatkiller commented:
> > These steps still only run on Security PRs
> > OS Query Cypress burns (these steps run new or changed tests multiple times before merge to ensure stability)
> > General Security Solution Cypress burns (these steps run new or changed tests multiple times before merge to ensure stability)
> 
> @oatkiller what is the reason behind not moving these as well? I believe it would be good to have them as well, since if I'm not mistaken, are just triggered when there is a change in one of the cypress specs, that would help us a lot to control/surface flakiness and they are not blockers since are soft failures.

Hey @MadameSheema. Thanks for the review. My original thought was this:

1. These steps only run changed or new Security Cypress tests
2. Therefore they only need to run on PRs that change those files

Alternatively we could run those steps on all PRs. However, for PRs that don't change Security's Cypress specs, they would just start a buildkite node, find no specs, and then terminate. 

| Topic | Burns always run | Burns only run when a subset of PRs are started |
| -- | --- | -- |
| Starting buildkite jobs | Extra buildkite jobs will be started. They'll find no specs to 'burn' and terminate. This could be wasteful | Buildkite jobs won't be started to run Cypress burn on PRs that don't even change Security's files |
| Configuring which files should trigger a cypress burn | If we always run cypress burn, there is no need for an allowlist of files in the buildkite code. There is less chance of a configuration mismatch between buildkite and cypress | The logic to determine if 'burn' will happen is split up. An initial filter runs in buildkite, and then a second filter runs in parallel.ts |

@MadameSheema @patrykkopycinski @jbudz do you have an opinion here? I think either approach is acceptable.

oatkiller commented on .buildkite/scripts/pipelines/pull_request/pipeline.ts:
```diff
@@ -76,7 +76,6 @@ const uploadPipeline = (pipelineContent: string | object) => {
       GITHUB_PR_LABELS.includes('ci:all-cypress-suites')
     ) {
       pipeline.push(getPipeline('.buildkite/pipelines/pull_request/security_solution.yml'));
```
These files now just run:
* cypress burn
* OpenAPI codegen
* Serverless OS Query tests

I don't think these steps match the `doAnyChangesMatch` call above. 

For example, changes to /^fleet_packages\.json/ probably shouldn't trigger those actions. Also, all PRs will run our tests now.

With that in mind, I think we need to re-evaluate these filters.

@xcrzx I believe you added the OpenAPI codegen step. Do you know when it needs to run?

@patrykkopycinski Should Serverless Osquery Cypress Tests run on all PRs now?

xcrzx replied:
> @xcrzx I believe you added the OpenAPI codegen step. Do you know when it needs to run?

We need to run it whenever there are changes to schema files (`*.schema.yaml`) or generated artifacts (`*.gen.ts`) in Security Solution. But as we're making the codegen script configurable, other teams might also start using it in their plugins.

patrykkopycinski replied:
yes, all Servereless Cypre Tests should should run on all PRs

jbudz commented:
>  do you have an opinion here? I think either approach is acceptable.

We don't have a great UX yet for soft failures.  When we had them enabled for serverless tests PR comments displaying `Build failed` but still allowing merges caused a lot of confusion.  I'm not opposed to further consideration and we can improve on the PR comment, but would it be okay to follow up with a separate issue?

oatkiller commented on .buildkite/pipelines/on_merge_unsupported_ftrs.yml:
```diff
@@ -49,66 +49,6 @@ steps:
         - exit_status: '*'
           limit: 1
 
-  - command: .buildkite/scripts/steps/functional/security_solution.sh
```
These were moved to `on_merge.yml`

oatkiller commented on .buildkite/pipelines/on_merge.yml:
```diff
@@ -114,6 +114,66 @@ steps:
         - exit_status: '*'
           limit: 1
 
+  - command: .buildkite/scripts/steps/functional/security_solution.sh
```
These were copied from `on_merge_unsupported_ftrs.yml`

jbudz reviewed (approved):
LGTM.  Timing wise, I would prefer waiting until the 8.11 branch is cut before merging.  If we're intentionally targeting this before, no problem

MadameSheema commented:
> @MadameSheema @patrykkopycinski @jbudz do you have an opinion here? I think either approach is acceptable.

@oatkiller Not a strong opinion on any of the above solutions if both are valid. My only thought is that if possible, would be great to minimize the number of files where our code lives, in this way is going to be easier to maintain the future. 

>> @oatkiller I believe we should do the same for the threat intelligence cypress tests.
>> @PhilippeOberti may you please confirm if those tests are still owned by us? Thanks!
> I'd say they are yes, as @lgestc and I are the ones who worked on them, and we're both in the Investigations' team

Are we going to move the threat intelligence tests as well?

> LGTM. Timing wise, I would prefer waiting until the 8.11 branch is cut before merging. If we're intentionally targeting this before, no problem.

Agree! It would be nice to merge it if possible after 8.11 branch cut to avoid possible issues.

oatkiller commented:
Hey all. I’m aiming to enable our Cypress e2e tests on all PRs. Can you help me find the right time to merge this? Tomorrow is feature freeze.

* Option 1: Merge before feature freeze. If we do this, then all of the PRs in the rush to feature-freeze will run our tests. That’s good news! The possible downside is that our tests may be flaky and we could end up blocking the build. On the other hand, we have a workflow for skipping flaky tests, so we have a way to deal with that.
* Option 2: Merge soon after FF. Once the 8.11 branch is cut, most changes going into 8.11 will already be in main. This will allow PRs to go into main without running our tests (as they do today) but it de-risks the release process of other teams.
* Option 3: Merge well after FF. Pick a date in the future (e.g. next week) and merge then.

cc: @paul-tavares @patrykkopycinski @kevinlog @eyalkraft @tehilashn @peluja1012 @marshallmain @yctercero @asnehalb @YulNaumenko @michaelolo24 @stephmilovic @MindyRS @MadameSheema @charlie-pichette @k-g-elastic @epixa @crowens @arpadkiraly

MadameSheema commented:
@oatkiller I vote for option 2, FF is happening tomorrow so the PR cut will arrive soon :)

kevinlog commented:
@oatkiller I agree with @MadameSheema on option 2.

oatkiller commented:
cc: @mchopda

paul-tavares commented:
Hey @oatkiller - thanks for the ping. I also think option 2 is best. 

As a side note: I wonder if some sort of "warning" to others that commit to Kibana should be sent out. I often encounter CI errors while the PR is pending that have nothing to do with the changes introduced by the PR (they are failures in other security solution test suites). I get around them by just keep merging upstream and they eventually generate a 💚  build, but this type of issue (unstable tests - maybe?) will likely a) frustrate others outside of security solution, b) cause an increase in "I need help with your failing tests" in our security team channels.

Just a heads up.

Commit:
Move Securitys Cypress tests to on_merge pipeline from on_merge_unsupported_ftrs

oatkiller commented:
> > > @MadameSheema @patrykkopycinski @jbudz do you have an opinion here? I think either approach is acceptable.
> > 
> > 
> > @oatkiller Not a strong opinion on any of the above solutions if both are valid. My only thought is that if possible, would be great to minimize the number of files where our code lives, in this way is going to be easier to maintain the future.
> > > > @oatkiller I believe we should do the same for the threat intelligence cypress tests.
> > > > @PhilippeOberti may you please confirm if those tests are still owned by us? Thanks!
> > > > I'd say they are yes, as @lgestc and I are the ones who worked on them, and we're both in the Investigations' team
> > 
> > 
> > Are we going to move the threat intelligence tests as well?
> > > LGTM. Timing wise, I would prefer waiting until the 8.11 branch is cut before merging. If we're intentionally targeting this before, no problem.
> > 
> > 
> > Agree! It would be nice to merge it if possible after 8.11 branch cut to avoid possible issues.
> 
> I will move the threat intelligence tests as well.

Thanks. I'll figure out a way to notify some more people.

Commit:
move the security solution pr pipeline into the base pipeline

Commit:
Move Threat Intel PR pipeline to the base.yml PR pipeline file

Commit:
move defend workflows pr pipeline into base.yml

Commit:
Move osquery cypress tests into base pr pipeline

and remove some branches from pipeline.ts that are no longer needed

Commit:
Update on_merge pipeline

ensure jobs in on_merge have the same labels and parallelism as the equivalent jobs in the pull_request pipeline
ensure all Security cypress tests, including serverless tests, are in on_merge

oatkiller commented on .buildkite/pipelines/pull_request/base.yml:
```diff
@@ -93,6 +93,136 @@ steps:
+          limit: 1
+
+  - command: .buildkite/scripts/steps/functional/defend_workflows_serverless.sh
+    label: 'Defend Workflows Cypress Tests on Serverless'
```
@patrykkopycinski is this right?

Commit:
Merge branch 'main' into enable-security-cypress-tests

oatkiller commented:
Update!:

Hey all, Thanks for the previous reviews. 

I have made a few changes:

* `burn` steps run on all PRs now. This seems to work fine.
* `codegen` steps for OpenAPI run on all PRs now
* Threat Intel steps run on all PRs

jbudz commented on .buildkite/pipelines/pull_request/base.yml:
```diff
@@ -93,6 +93,136 @@ steps:
+      automatic: false
+    soft_fail: true
+
+  - command: .buildkite/scripts/steps/code_generation/security_solution_codegen.sh
```
This has a short [runtime](https://buildkite.com/elastic/kibana-pull-request/builds/165962#018b0550-718a-4c81-b720-f81a46e17642) can we add it to `.buildkite/scripts/steps/checks.sh` instead?  It had its own step before due to it running conditionally.

Okay with addressing in a followup.

oatkiller replied:
@jbudz thanks for the tip. Here is a follow up PR with that change: https://github.com/elastic/kibana/pull/168260

PhilippeOberti reviewed (approved):
LGTM for the Protections Experience team!

nkhristinin reviewed (approved):
LGTM!
